library/hashmap.lisp

```diff
@@ -0,0 +1,523 @@
+      ((== len 2)
+       ;; If we get 2-entry array and the remaining branch is Leaf,
+       ;; we can eliminate the Tree node.
+       ;; In this case, pos is 0 or 1, so remaining node pos is (- 1 pos)
+       (let ((pos (index->pos mask index))
+             (remaining (arr:aref array (- 1 pos))))
+         (match remaining
+           ((Leaf entry) (Some (Leaf entry))) ;drop tree
+           (_ (Some (newtree))))))
+      (True (Some (newtree)))))
+
+  (define (new-tree-1 depth entry)
+    "Create a new tree with ENTRY as a sole branch.  The caller will add a
+new entry to the returned node."
+    (let ((ind1 (trie-index (hbits (entry-key entry)) depth)))
+      (Tree (index->mask ind1) (arr:make 1 (Leaf entry)))))
+
+  (define (new-tree-2 depth entry1 entry2)
+    "Create a new tree with ENTRY1 and ENTRY2 as branches.  The caller will add
+a new entry."
+    (let ((ind1 (trie-index (hbits (entry-key entry1)) depth))
+          (ind2 (trie-index (hbits (entry-key entry2)) depth)))
+      (if (== ind1 ind2)
+          (Tree (index->mask ind1) (arr:make 1 (Bud entry1 entry2)))
+          (let ((arr (arr:make 2 (Chain Nil)))
+                (mask (bits:or (index->mask ind1) (index->mask ind2))))
+            (cond ((< ind1 ind2)
+                   (arr:set! arr 0 (Leaf entry1))
+                   (arr:set! arr 1 (Leaf entry2)))
+                  (True
+                   (arr:set! arr 0 (Leaf entry2))
+                   (arr:set! arr 1 (Leaf entry1))))
+            (Tree mask arr)))))
+
+  ;; API
+  (declare empty (HashMap :k :v))
+  (define empty
+    "An empty HashMap"
+    (HashMap (Chain Nil)))
+
+  ;; API
+  (declare empty? (HashMap :k :v -> Boolean))
+  (define (empty? hm)
+    "Returns True if a hashmap HM is empty, False if not."
+    (match (.root hm)
+      ((Chain (Nil)) True)
+      (_ False)))
+
+  ;; API
+  (declare count (HashMap :k :v -> Integer))
+  (define (count hm)
+    "Returns the number of entries in HM."
+    (into
+     (rec walk ((node (.root hm)))
+       (match node
+         ((Leaf _) 1)
+         ((Bud _ _) 2)
+         ((Chain lis) (list:length lis))
+         ((Tree _ array)
+          (fold (fn (sum elt) (+ sum (walk elt))) 0 array))))))
+
+  ;; API
+  (declare get (Hash :k => HashMap :k :v -> :k -> Optional :v))
+  (define (get hm key)
+    "Returns a value associated with KEY in the hashmap HM."
+    (let hb = (hbits key))
+    (rec search ((depth 0)
+                 (node (.root hm)))
+      (match node
+        ((Leaf (HmEntry k v))
+         (if (== key k)
+             (Some v)
+             None))
+        ((Bud (HmEntry k1 v1) (HmEntry k2 v2))
+         (cond ((== key k1) (Some v1))
+               ((== key k2) (Some v2))
+               (True None)))
+        ((Chain entries)
+         (match (iter:find! (fn (e) (== key (entry-key e)))
+                            (iter:into-iter entries))
+           ((None) None)
+           ((Some e) (Some (entry-value e)))))
```

Comment:
style nit. Consider `map`

``` lisp
(map entry-value (iter:find! (fn (e) ....))
```

or `do` 

``` lisp
(do 
  (e <- (iter:find! (fn (e) ...))
  (pure (entry-value e)))
```

Although, either one is possibly less clear than `match`